### PR TITLE
FindPointsGSLIB on GPUs

### PR DESCRIPTION
**Pull request by kmittal2**
| PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4467](https://github.com/mfem/mfem/pull/4467) | @kmittal2 | @tzanio | @helloworld922 + @vladotomov + @SchmidtMat | 9/6/24 | 12/9/24 | 12/17/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by helloworld922** [comment]
`fem/CMakeLists.txt` is missing the source files in `fem/gslib` so cmake builds fail at link time.

File: fem/gslib.cpp [hunk 1]
```diff
@@ -853,6 +1915,247 @@ void FindPointsGSLIB::MapRefPosAndElemIndices()
...
+      MPI_Barrier(gsl_comm->c);
+      interp_vals.HostRead();
+
+      // Now the interpolated values need to be sent back component wise
+      int Offset = interp_vals.Size()/ncomp;
+      for (int i = 0; i < ncomp; i++)
+      {
+         spt = (evalSrcPt_t *)src.ptr;
+         array_init(evalOutPt_t, &outpt, n);
+         outpt.n = n;
```
Comment (helloworld922) [inline fem/gslib.cpp:2129]: I think this can be moved outside the i loop?

File: fem/gslib.cpp [hunk 2]
```diff
@@ -853,6 +1915,247 @@ void FindPointsGSLIB::MapRefPosAndElemIndices()
...
+         opt = (evalOutPt_t *)outpt.ptr;
+         for (int index = 0; index < outpt.n; index++)
+         {
+            int idx = ordering == Ordering::byNODES ?
+                      opt->index + i*points_cnt :
+                      opt->index*ncomp + i;
+            field_out(idx) = opt->out;
+            ++opt;
+         }
+         array_free(&outpt);
```
Comment (helloworld922) [inline fem/gslib.cpp:2152]: can move outside the i loop?

File: fem/gslib/findpts_local_2.cpp [hunk 3]
```diff
@@ -0,0 +1,1246 @@
...
+                                     const int pN = 0)
+{
+#define MAX_CONST(a, b) (((a) > (b)) ? (a) : (b))
+   const int MD1 = T_D1D ? T_D1D : 14;
+   const int D1D = T_D1D ? T_D1D : pN;
+   const int p_NE = D1D*D1D;
+   const int p_NEL = nel*p_NE;
+   MFEM_VERIFY(MD1 <= 14,"Increase Max allowable polynomial order.");
+   MFEM_VERIFY(D1D != 0, "Polynomial order not specified.");
+   const int nThreads = 32;
```
Comment (helloworld922) [inline fem/gslib/findpts_local_2.cpp:717]: Should this be a tune-able parameter?

File: fem/gslib/findpts_local_3.cpp [hunk 4]
```diff
@@ -0,0 +1,1922 @@
...
+namespace mfem
+{
+
+#define CODE_INTERNAL 0
+#define CODE_BORDER 1
+#define CODE_NOT_FOUND 2
+#define DIM 3
+#define DIM2 DIM*DIM
+#define pMax 10
+#define nThreads 32
```
Comment (helloworld922) [inline fem/gslib/findpts_local_3.cpp:27]: should this be a tune-able parameter?

**Review by helloworld922 — CHANGES_REQUESTED** [review]
See comments. The first three comments were required to get mfem+gslib to build for me using CMake (latter two of these required for makefile build).

The rest are discussion points about the implementation and changes won't affect final results outside of performance, if this is even intended to be performance critical. I imagine that performance might be desired if the goal is to get this to run on device?

There were a few other points where I noticed some allocations which don't change size inside of loops (gslib.cpp:2481, gslib.cpp:2587+gslib.cpp:2590) but these were in the original version of the code. Moving these outside the loop might be worthwhile to do either in this merge request or a separate one if the goal is improve the performance of gslib.

File: fem/gslib/findpts_local_3.cpp [hunk 5]
```diff
@@ -0,0 +1,1922 @@
...
+void FindPointsGSLIB::FindPointsLocal3(const Vector &point_pos,
+                                       int point_pos_ordering,
+                                       Array<unsigned int> &code,
+                                       Array<unsigned int> &elem,
+                                       Vector &ref,
+                                       Vector &dist,
+                                       int npt)
+{
+   SWkernel.Clear();
+   SWkernel.Start();
```
[the remark this answers is not in the document]
Comment (kmittal2) [reply fem/gslib/findpts_local_3.cpp]: Good catch!

File: fem/gslib.hpp [hunk 6]
```diff
@@ -91,10 +89,26 @@ class FindPointsGSLIB
...
+      int local_hash_size, dof1d, dof1d_sol, h_o_size, h_nx;
+      double tol;
+      struct gslib::crystal *cr;
+      struct gslib::hash_data_3 *hash3;
+      struct gslib::hash_data_2 *hash2;
+      mutable Vector bb, wtend, gll1d, lagcoeff, gll1d_sol, lagcoeff_sol;
+
+      mutable Array<unsigned int> loc_hash_offset;
+      mutable Vector loc_hash_min, loc_hash_fac;
+   } DEV;
```
[the remark this answers is not in the document]
Comment (kmittal2) [reply fem/gslib.hpp]: I added some documentation for the different tolerances.. I have kept Device specific things in a different struct for clarity and following what we do for the [TMOPIntegrator class](https://github.com/mfem/mfem/blob/master/fem/tmop.hpp#L1870-L1883).

File: fem/gslib.cpp (hunk 1 above)
Comment (kmittal2) [reply fem/gslib.cpp:2131]: The call to `sarray_transfer` redistributes the contents in parallel so we need to keep this inside the loop.

File: fem/gslib/findpts_local_2.cpp (hunk 3 above)
Comment (kmittal2) [reply fem/gslib/findpts_local_2.cpp:717]: yeah I could define it as `MD1*DIM` inside the kernel?

**Review by helloworld922 — APPROVED** [review]

**Comment by tzanio** [comment]
The code style check is failing

**Review by vladotomov — CHANGES_REQUESTED** [review]
- CHANGELOG (explain the GPU support is fully MFEM-native).
- merge master (produces conflicts).
- explain better what meshes are produced by `GetBoundingBoxMesh` (if it needs to stay `public`) (with doxygen).

**Review by vladotomov — APPROVED** [review]
Very impressive work Ketan!

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
Some of the nightly tests failed when trying to install the library with this branch, so I added GitHub CI tests to check that the install targets work.

**Comment by v-dobrev** [comment]
Re-merged in `next` for testing...

**Comment by v-dobrev** [comment]
@kmittal2, I was building this PR with GSLIB 1.0.9 enabled and compiler flags like the ones we use for catching warnings in CI: `-Werror -pedantic -Wall -Wshadow` (on Mac or with Clang) -- this generated some errors (from warnings) that may be worth fixing.

**Comment by vladotomov** [comment]
I think I cleaned them.

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by v-dobrev** [comment]
> I think I cleaned them.

There are some more errors (from warnings) when you build everything with `make all` -- these are not new but it's good to fix them.

**Comment by tzanio** [comment]
@mstees, does this work for you?

**Comment by vladotomov** [comment]
ok I cleaned `make all` too

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by artv3** [comment]
Hi all, just wanted to follow up here. Is this PR ready to be merged?

**Comment by vladotomov** [comment]
@artv3 we're still investigating the errors reported by @mstees

**Review by mstees — APPROVED** [review]

**Comment by vladotomov** [comment]
The issues were resolved by updating the ROCm version, thank you @camierjs and @mstees!

**Review by tzanio — APPROVED** [review]
Thank you @kmittal2, @v-dobrev and @camierjs !

**Pull request merged**